packages/language-server/src/edit/edit/transactionalEdit.ts

```diff
@@ -0,0 +1,340 @@
+     * Contains additional edits for add edit engines
+     */
+    readonly initialTextEdits: TextEdit[] = [];
+
+    /**
+     * Element type lookup of the initial diagram
+     * Used to check if predictions are still valid
+     */
+    private readonly initialDiagramElementLookup: Map<string, string> = new Map();
+
+    /**
+     * Creates a new transactional edit
+     *
+     * @param action the action this edit is based on
+     * @param diagram the diagram affected by the action
+     * @param registry the registry to use
+     */
+    constructor(
+        private readonly action: TransactionalAction,
+        private readonly diagram: Diagram,
+        private readonly registry: EditHandlerRegistry
+    ) {
+        this.transactionId = action.transactionId;
+        this.initialSequenceNumber = action.sequenceNumber;
+        for (const element of Object.values(diagram.currentDiagram!.elementLookup)) {
+            this.initialDiagramElementLookup.set(element.id, element.type);
+        }
+        this.generateEngines();
+    }
+
+    /**
+     * Applies the action and generates the text document edit based on it
+     *
+     * @param action the action to apply
+     * @returns the text document edit
+     */
+    async applyAction(action: TransactionalAction): Promise<TextDocumentEdit> {
+        const edits: TextEdit[] = [];
+        if (action.sequenceNumber === this.initialSequenceNumber) {
+            edits.push(...this.initialTextEdits);
+        }
+        const textDocument = this.diagram.document;
+        for (const engine of this.engines) {
+            const text = await engine.apply(action.edits.map((edit) => edit.values));
+            edits.push({
+                range: Range.create(textDocument.positionAt(engine.start), textDocument.positionAt(engine.end)),
+                newText: text
+            });
+        }
+        return TextDocumentEdit.create(
+            VersionedTextDocumentIdentifier.create(textDocument.uri, textDocument.version),
+            edits
+        );
+    }
+
+    /**
+     * Update the engines based on the text document changes.
+     * Updates indices.
+     *
+     * @param changes the changes to the text document
+     */
+    updateEngines(changes: TextDocumentContentChangeEvent[]): void {
+        const textDocument = this.diagram.document;
+        for (const change of changes) {
+            if (!TextDocumentContentChangeEvent.isIncremental(change)) {
+                break;
+            }
+            const start = textDocument.offsetAt(change.range.start);
+            const end = textDocument.offsetAt(change.range.end);
+            const delta = change.text.length - (end - start);
+            for (const entry of this.engines) {
+                if (entry.start > end) {
+                    entry.start += delta;
+                    entry.end += delta;
+                } else if (entry.start >= start) {
+                    entry.end += delta;
+                }
+            }
```

Comment:
there are 4 cases
- the change is completely before the range: fix the start offset of the range, as it now starts later/earlier
- the change is the exact same range (it was a replacement): fix the length of the range
- the change is completely after the range: do nothing
- the change overlaps in any way than "identical": this is an error case which should never happen, if you want I can add a check for this (I was to lazy to do this in the past because special care is needed to get this one right and not produce false positives)